### PR TITLE
Fix usage of process_vm_readv in createdump

### PR DESCRIPTION
**Pull request by hoyosjs**
This PR tries to use `process_vm_readv` and falls back to using `pread` on `/proc/mem` if the first fails with `EPERM`. This is needed as docker 19.03 on kernels newer than 4.8 allows to use `ptrace` as long as yama settings don't restrict it, but it does not allow usage of `process_vm_*` syscalls. These were added very recently to the allow-list but we need a mechanism to allow customers to collect dumps in their containerized environments.

Fixes https://github.com/dotnet/diagnostics/issues/2098 in main.

### CODE REVIEW

File: src/coreclr/debug/createdump/crashinfounix.cpp [hunk 1]
```diff
@@ -334,17 +332,31 @@ CrashInfo::ReadProcessMemory(void* address, void* buffer, size_t size, size_t* r
...
+        // performance optimization.
+        m_canUseProcVmReadSyscall = false;
+        assert(m_fd != -1);
+        *read = pread64(m_fd, buffer, size, (off64_t)address);
+    }
+
     if (*read == (size_t)-1)
     {
+        int readErrno = errno;
+        TRACE("ReadProcessMemory FAILED, addr: %" PRIA PRIx ", size: %zu, ERRNO %d: %s\n", address, size, readErrno, strerror(readErrno));
```
Comment (mikem8361) [inline src/coreclr/debug/createdump/crashinfounix.cpp:359]: Doesn't this TRACE make the trace logging too noisy?

**Review by mikem8361 — APPROVED** [review]

**Comment by tommcdon** [comment]
fyi @shirhatti

File: src/coreclr/debug/createdump/crashinfounix.cpp (hunk 1 above)
Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfounix.cpp:359]: You can turn into a TRACE_VERBOSE which is normally a nop.

Comment (hoyosjs) [reply src/coreclr/debug/createdump/crashinfounix.cpp:359]:
```suggestion
        TRACE_VERBOSE("ReadProcessMemory FAILED, addr: %" PRIA PRIx ", size: %zu, ERRNO %d: %s\n", address, size, readErrno, strerror(readErrno));
```

Comment (mikem8361) [reply src/coreclr/debug/createdump/crashinfounix.cpp]: Looks good.

**Comment by saul** [comment]
Thanks for the fix. Can we expect this to be backported to 5.0?

**Comment by hoyosjs** [comment]
@saul I've opened the backport PR. Will give the change some bake time against the tests and ask ship room for permission to add to servicing in 5.0.

**Pull request merged**